File: php/phpfind/tests/FileUtilTest.php
<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;

use phpfind\FileUtil;

class FileUtilTest extends TestCase
{
    /***************************************************************************
     * expand_path tests
     **************************************************************************/
    public function test_expand_path_tilde(): void
    {
        $path = '~';
        $expected = $_SERVER['HOME'];
        $this->assertEquals($expected, FileUtil::expand_path($path));
    }

    public function test_expand_path_with_tilde(): void
    {
        $path = '~/src/xfind';
        $expected = FileUtil::join_paths($_SERVER['HOME'], 'src/xfind');
        $this->assertEquals($expected, FileUtil::expand_path($path));
    }

    public function test_expand_path_with_tilde_and_name(): void
    {
        $path = '~cary/src/xfind';
        $expected = FileUtil::join_paths($_SERVER['HOME'], 'src/xfind');
        $this->assertEquals($expected, FileUtil::expand_path($path));
    }

    /***************************************************************************
     * get_extension tests
     **************************************************************************/
    public function test_get_extension_has_txt_extension(): void
    {
        $file_name = 'filename.txt';
        $this->assertEquals("txt", FileUtil::get_extension($file_name));
    }

    public function test_get_extension_missing_extension(): void
    {
        $file_name = 'filename.';
        $this->assertEquals("", FileUtil::get_extension($file_name));
    }

    public function test_get_extension_no_extension(): void
    {
        $file_name = 'filename';
        $this->assertEquals("", FileUtil::get_extension($file_name));
    }

    public function test_get_extension_hidden_txt_extension(): void
    {
        $file_name = '.filename.txt';
        $this->assertEquals("txt", FileUtil::get_extension($file_name));
    }

    public function test_get_extension_hidden_missing_extension(): void
    {
        $file_name = '.filename.';
        $this->assertEquals("", FileUtil::get_extension($file_name));
    }

    public function test_get_extension_hidden_no_extension(): void
    {
        $file_name = '.filename';
        $this->assertEquals("", FileUtil::get_extension($file_name));
    }

    /***************************************************************************
     * is_dot_dir tests
     **************************************************************************/
    public function test_is_dot_dir_single_dot(): void
    {
        $file_name = '.';
        $this->assertTrue(FileUtil::is_dot_dir($file_name));
    }

    public function test_is_dot_dir_double_dot(): void
    {
        $file_name = '..';
        $this->assertTrue(FileUtil::is_dot_dir($file_name));
    }

    public function test_is_dot_dir_non_dot_dir(): void
    {
        $file_name = '.git';
        $this->assertFalse(FileUtil::is_dot_dir($file_name));
    }

    /***************************************************************************
     * is_hidden tests
     **************************************************************************/
    public function test_is_hidden_hidden_file(): void
    {
        $file_name = '.filename.txt';
        $this->assertTrue(FileUtil::is_hidden($file_name));
    }

    public function test_is_hidden_not_hidden_file(): void
    {
        $file_name = 'filename.txt';
        $this->assertFalse(FileUtil::is_hidden($file_name));
    }

    public function test_is_hidden_single_dot(): void
    {
        $file_name = '.';
        $this->assertFalse(FileUtil::is_hidden($file_name));
    }

    public function test_is_hidden_double_dot(): void
    {
        $file_name = '..';
        $this->assertFalse(FileUtil::is_hidden($file_name));
    }

    /***************************************************************************
     * join_path tests
     **************************************************************************/
    public function test_join_path_forward_slashes(): void
    {
        // $path = '/path/to/nowhere';
        $path = DIRECTORY_SEPARATOR . implode(DIRECTORY_SEPARATOR, ['path', 'to', 'nowhere']);
        $file = 'nowhere.txt';
        $joined = FileUtil::join_paths($path, $file);
        $this->assertEquals('/path/to/nowhere/nowhere.txt', $joined);
    }

    public function test_join_path_no_slashes(): void
    {
        $path = 'nowhere';
        $file = 'nowhere.txt';
        $expected = implode(DIRECTORY_SEPARATOR, ['nowhere', 'nowhere.txt']);
        $joined = FileUtil::join_paths($path, $file);
        $this->assertEquals($expected, $joined);
    }

    /***************************************************************************
     * normalize_path tests
     **************************************************************************/
    public function test_normalize_path_no_trailing_slash(): void
    {
        // $path = '/path/to/nowhere';
        $path = DIRECTORY_SEPARATOR . implode(DIRECTORY_SEPARATOR, ['path', 'to', 'nowhere']);
        $this->assertEquals(FileUtil::normalize_path($path), $path);
    }

    public function test_normalize_path_trailing_slash(): void
    {
        // $path = '/path/to/nowhere/';
        $path = DIRECTORY_SEPARATOR . implode(DIRECTORY_SEPARATOR, ['path', 'to', 'nowhere']);
        $this->assertEquals(
            $path,
            FileUtil::normalize_path($path . DIRECTORY_SEPARATOR)
        );
    }

    /***************************************************************************
     * split_path tests
     **************************************************************************/
    public function test_split_path_no_sep(): void
    {
        $path = 'nowhere';
        $split = FileUtil::split_path($path);
        $this->assertCount(1, $split);
        $this->assertEquals('nowhere', $split[0]);
    }

    public function test_split_path_slashes(): void
    {
        // $path = '/path/to/nowhere';
        $path = DIRECTORY_SEPARATOR . implode(DIRECTORY_SEPARATOR, ['path', 'to', 'nowhere']);
        $split = FileUtil::split_path($path);
        $this->assertCount(4, $split);
        $this->assertEquals('', $split[0]);
        $this->assertEquals('path', $split[1]);
        $this->assertEquals('to', $split[2]);
        $this->assertEquals('nowhere', $split[3]);
    }
}

File: perl/plfind/t/fileutil_test.pl
#!/usr/bin/perl -w
#
# fileutil_test.pl
#
#
use strict;
use warnings;

use Cwd 'abs_path';
use File::Basename;

my $lib_path;

BEGIN {
    $lib_path = dirname(dirname(abs_path($0))) . '/lib';
    # print "lib_path: $lib_path\n";
    unshift @INC, $lib_path;
}

use Test::Simple tests => 16;

use plfind::FileUtil;

################################################################################
# expand_path tests
################################################################################
sub test_expand_path_tilde {
    my $path = '~';
    my $expected = $ENV{HOME};
    my $expanded = plfind::FileUtil::expand_path($path);
    ok($expanded eq $expected, "$path expands to home path");
}

sub test_expand_path_with_tilde {
    my $path = '~/src/xfind';
    my $expected = $ENV{HOME} . '/src/xfind';
    my $expanded = plfind::FileUtil::expand_path($path);
    ok($expanded eq $expected, "$path expands to start with home path");
}

sub test_expand_path_with_tilde_and_name {
    my $path = '~cary/src/xfind';
    my $expected = $ENV{HOME} . '/src/xfind';
    my $expanded = plfind::FileUtil::expand_path($path);
    ok($expanded eq $expected, "$path expands to start with home path");
}

################################################################################
# get_extension tests
################################################################################
sub test_get_extension_has_txt_extension {
    my $file_name = 'filename.txt';
    my $ext = plfind::FileUtil::get_extension($file_name);
    ok($ext eq "txt", "$file_name has extension txt");
}

sub test_get_extension_missing_extension {
    my $file_name = 'filename.';
    my $ext = plfind::FileUtil::get_extension($file_name);
    ok($ext eq "", "$file_name has missing extension");
}

sub test_get_extension_no_extension {
    my $file_name = 'filename';
    my $ext = plfind::FileUtil::get_extension($file_name);
    ok($ext eq "", "$file_name has no extension");
}

sub test_get_extension_hidden_txt_extension {
    my $file_name = '.filename.txt';
    my $ext = plfind::FileUtil::get_extension($file_name);
    ok($ext eq "txt", "$file_name has extension txt");
}

sub test_get_extension_hidden_missing_extension {
    my $file_name = '.filename.';
    my $ext = plfind::FileUtil::get_extension($file_name);
    ok($ext eq "", "$file_name has missing extension");
}

sub test_get_extension_hidden_no_extension {
    my $file_name = '.filename';
    my $ext = plfind::FileUtil::get_extension($file_name);
    ok($ext eq "", "$file_name has no extension");
}

################################################################################
# is_dot_dir tests
################################################################################
sub test_is_dot_dir_single_dot {
    my $file_name = '.';
    my $ok = plfind::FileUtil::is_dot_dir($file_name);
    ok($ok > 0, "$file_name is dot dir");
}

sub test_is_dot_dir_double_dot {
    my $file_name = '..';
    my $ok = plfind::FileUtil::is_dot_dir($file_name);
    ok($ok > 0, "$file_name is dot dir");
}

sub test_is_dot_dir_non_dot_dir {
    my $file_name = '.git';
    my $ok = plfind::FileUtil::is_dot_dir($file_name);
    ok($ok == 0, "$file_name is not dot dir");
}

################################################################################
# is_hidden tests
################################################################################
sub test_is_hidden_hidden_file {
    my $file_name = '.filename.txt';
    my $ok = plfind::FileUtil::is_hidden($file_name);
    ok($ok > 0, "$file_name is hidden file");
}

sub test_is_hidden_not_hidden_file {
    my $file_name = 'filename.txt';
    my $ok = plfind::FileUtil::is_hidden($file_name);
    ok($ok == 0, "$file_name is not hidden file");
}

sub test_is_hidden_single_dot {
    my $file_name = '.';
    my $ok = plfind::FileUtil::is_hidden($file_name);
    ok($ok == 0, "$file_name is not hidden file");
}

sub test_is_hidden_double_dot {
    my $file_name = '..';
    my $ok = plfind::FileUtil::is_hidden($file_name);
    ok($ok == 0, "$file_name is not hidden file");
}

################################################################################
# main
################################################################################
sub main {
    test_expand_path_tilde();
    test_expand_path_with_tilde();
    test_expand_path_with_tilde_and_name();
    test_get_extension_has_txt_extension();
    test_get_extension_missing_extension();
    test_get_extension_no_extension();
    test_get_extension_hidden_txt_extension();
    test_get_extension_hidden_missing_extension();
    test_get_extension_hidden_no_extension();
    test_is_dot_dir_single_dot();
    test_is_dot_dir_double_dot();
    test_is_dot_dir_non_dot_dir();
    test_is_hidden_hidden_file();
    test_is_hidden_not_hidden_file();
    test_is_hidden_single_dot();
    test_is_hidden_double_dot();
}

main();
